src/cli/expand-patterns.js

```diff
@@ -0,0 +1,234 @@
+const flat = require("lodash/flatten");
+
+/** @typedef {import('./util').Context} Context */
+
+/**
+ * @param {Context} context
+ */
+function* expandPatterns(context) {
+  const cwd = process.cwd();
+  const seen = Object.create(null);
+  let noResults = true;
+
+  for (const pathOrError of expandPatternsInternal(context)) {
+    noResults = false;
+    if (typeof pathOrError !== "string") {
+      yield pathOrError;
+      continue;
+    }
+
+    const relativePath = path.relative(cwd, pathOrError);
+
+    // filter out duplicates
+    if (relativePath in seen) {
+      continue;
+    }
+
+    seen[relativePath] = true;
+    yield relativePath;
+  }
+
+  if (noResults) {
+    // If there was no files and no other errors, let's yield a general error.
+    yield {
+      error: `No matching files. Patterns: ${context.filePatterns.join(" ")}`
+    };
+  }
+}
+
+// TODO: use `fast-glob` directly, without `globby`
+const baseGlobbyOptions = {
+  dot: true,
+  expandDirectories: false,
+  absolute: true
+};
+
+/**
+ * @param {Context} context
+ */
+function* expandPatternsInternal(context) {
+  // Ignores files in version control systems directories and `node_modules`
+  const ignoredDirectories = {
+    ".git": true,
+    ".svn": true,
+    ".hg": true,
+    node_modules: context.argv["with-node-modules"] !== true
+  };
+
+  const globbyOptions = {
+    ...baseGlobbyOptions,
+    ignore: Object.keys(ignoredDirectories)
+      .filter(dir => ignoredDirectories[dir])
+      .map(dir => "**/" + dir)
+  };
+
+  let supportedFilesGlob;
+  const cwd = process.cwd();
+
+  /**
+   * @type {Array<{
+   *  type: 'file' | 'dir' | 'glob';
+   *  absolutePath?: string;
+   *  glob?: string;
+   *  input: string;
+   * }>}
+   */
+  const entries = [];
+
+  for (const pattern of context.filePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+
+    if (containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories)) {
+      continue;
+    }
+
+    const stat = statSafeSync(absolutePath);
+    if (stat) {
+      if (stat.isFile()) {
+        entries.push({ type: "file", absolutePath, input: pattern });
+      } else if (stat.isDirectory()) {
+        entries.push({
+          type: "dir",
+          absolutePath,
+          glob: escapePathForGlob(pattern) + "/" + getSupportedFilesGlob(),
+          input: pattern
+        });
+      }
+    } else if (pattern[0] === "!") {
+      // convert negative patterns to `ignore` entries
+      globbyOptions.ignore.push(fixWindowsSlashes(pattern.slice(1)));
+    } else {
+      entries.push({
+        type: "glob",
+        glob: fixWindowsSlashes(pattern),
+        input: pattern
+      });
+    }
+  }
+
+  for (const { type, absolutePath, glob, input } of entries) {
+    switch (type) {
+      case "file":
+        yield absolutePath;
+        continue;
+
+      case "dir": {
+        let result;
+        try {
+          result = globby.sync(glob, globbyOptions);
```

Comment:
Looks here we have `expandDirectories: false`, so nested directories from `prettier dir` is not handled, right? I am on mobile, so can't test :smile:

Reply:
They're handled because a glob like this is composed for dirs: `<dir name>/**/<all supported extension and well-known filenames>`

When I implement [this suggestion](https://github.com/prettier/prettier/pull/7660#discussion_r383698135), it'll change to `<dir name>/**/*` with subsequent filtering based on parser inference.